staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go

```diff
@@ -121,7 +122,14 @@ func (j *JSONPath) FindResults(data interface{}) ([][]reflect.Value, error) {
 // PrintResults writes the results into writer
 func (j *JSONPath) PrintResults(wr io.Writer, results []reflect.Value) error {
 	for i, r := range results {
-		text, err := j.evalToText(r)
+		var text []byte
+		var err error
+		if r.Kind() == reflect.Interface && (r.Elem().Kind() == reflect.Map ||
+			r.Elem().Kind() == reflect.Slice || r.Elem().Kind() == reflect.Struct) {
```

Comment:
Do you want this change to apply for slice and struct?  I thought the issue was only about map.  I guess I would need to see some unit tests around those types to understand how the output would be different.

Reply:
Based on the comments in #16707 (@thockin last comment and others) indicated it'd be favorable to address all non-primitive types which could be represented by json.  Go-style output of structs does not contain attribute/key names and slices are printed without comma separators, which makes passing these outputs to other tools more difficult.

Reply:
The current test failure is an example jsonpath output where the slice was previously output is Go-style instead of json.

Expected was the old output, and actual is the proper json representation of the slice.
```
--- FAIL: TestTokenOutput (0.00s)
    --- PASS: TestTokenOutput/JSON_output (0.00s)
    --- PASS: TestTokenOutput/YAML_output (0.00s)
    --- PASS: TestTokenOutput/Go_template_output (0.00s)
    --- PASS: TestTokenOutput/text_output (0.00s)
    --- FAIL: TestTokenOutput/jsonpath_output (0.00s)
        token_test.go:449: failed TestTokenOutput:
            
            expected:
            abcdef.1234567890123456 [system:bootstrappers:kubeadm:default-node-token]
            
            actual:
            abcdef.1234567890123456 ["system:bootstrappers:kubeadm:default-node-token"]
```

Reply:
Can you add a couple unit tests to cover the slice and struct scenarios.
Something like this might work to cover both (slice with multiple structs):

```
{range .items[1]}{.status.addresses} {end}
```

I'm just thinking it will be good to have this to show the expected output and prevent any future regression.

Reply:
I've added several test cases for each primitive type.  While adding test cases, I resolved some missing/broken filter behavior that was skipping certain types / unstable sorting.